.github/workflows/test.yml

```diff
@@ -0,0 +1,111 @@
+# This is a GitHub workflow defining a set of jobs with a set of steps.
+# ref: https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions
+#
+name: Tests
+
+on:
+  pull_request:
+  push:
+
+defaults:
+  run:
+    shell: bash
+
+jobs:
+  lint:
+    runs-on: ubuntu-20.04
+    timeout-minutes: 2
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: actions/setup-python@v2
+        with:
+          python-version: 3.8
+      - uses: pre-commit/action@v2.0.0
+
+  # Run tests
+  test:
+    runs-on: ubuntu-20.04
+    timeout-minutes: 10
+
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - python: 3.8
+            test: internal-ssl
+          - python: 3.6
+            jupyterhub: "1.0"
+          - python: 3.7
+            jupyterhub: 1.1
+          - python: 3.8
+            jupyterhub: 1.2
+          - python: 3.9
+            jupyterhub: 1.3
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: setup docker swarm
+        run: docker swarm init
+
+      - name: Install Python ${{ matrix.python }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python }}
+
+      # preserve pip cache to speed up installation
+      - name: Cache pip
+        uses: actions/cache@v2
+        with:
+          path: ~/.cache/pip
+          # Look to see if there is a cache hit for the corresponding requirements file
+          key: ${{ runner.os }}-pip-${{ hashFiles('*requirements.txt') }}
+          restore-keys: |
+            ${{ runner.os }}-pip-
```

Comment:
I have bumped over this issue https://github.com/pypa/pip/issues/4537 when trying to make sense on why the TLJH unit tests [are failing](https://app.circleci.com/pipelines/github/jupyterhub/the-littlest-jupyterhub/219/workflows/5088d3ab-cc2e-4c60-909f-a72e15e9c37d/jobs/2740).
A possible solution I had in mind was to only keep the most restrictive key and don't have a restore one. Basically if anything changes in the requirement.txt, then just install everything.

I can't say if it's the case here too because I still don't think I fully understand the underlying issue :( But I thought it's worth knowing about this.